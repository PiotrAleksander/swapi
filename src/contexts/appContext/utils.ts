import { ResourceType } from "../../types/enums/ResourceTypeEnum";
import api from "../../services/swapi";
import { Resource } from "../../types/Resource";
import {
  isPeopleResource,
  isStarshipResource,
  isNotFound
} from "../../types/typeguards";
import { Dispatch, SetStateAction } from "react";

export const fetchResources = async (resourceType: ResourceType) => {
  let firstIndex = Math.round(Math.random() * 10) + 1;
  let secondIndex = Math.round(Math.random() * 10) + 1;

  if (firstIndex === secondIndex) secondIndex += 1;

  const firstResourcePromise = api.fetchResource(resourceType, firstIndex);
  const secondResourcePromise = api.fetchResource(resourceType, secondIndex);
  const [firstResource, secondResource] = await Promise.all([
    firstResourcePromise,
    secondResourcePromise
  ]);
  return [firstResource, secondResource];
};

export const setWinnerAndCounter = (
  resources: Array<Resource>,
  setCounter: Dispatch<SetStateAction<number[]>>
): Array<Resource> => {
  let firstResource = resources[0];
  let secondResource = resources[1];
  if (isNotFound(firstResource) || isNotFound(secondResource)) {
    return resources;
  }

  if (isPeopleResource(firstResource) && isPeopleResource(secondResource)) {
    firstResource.winner =
      parseInt(firstResource.mass) > parseInt(secondResource.mass);
    secondResource.winner =
      parseInt(secondResource.mass) > parseInt(firstResource.mass);
  }

  if (isStarshipResource(firstResource) && isStarshipResource(secondResource)) {
    firstResource.winner =
      parseInt(firstResource.crew) > parseInt(secondResource.crew);
    secondResource.winner =
      parseInt(secondResource.crew) > parseInt(firstResource.crew);
  }

  if (firstResource.winner) {
    setCounter(prevCounter => [prevCounter[0] + 1, prevCounter[1]]);
  }

  if (secondResource.winner) {
    setCounter(prevCounter => [prevCounter[0], prevCounter[1] + 1]);
  }

  return [firstResource, secondResource];
};
