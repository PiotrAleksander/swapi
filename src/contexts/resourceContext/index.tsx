import React, { createContext } from "react";
import { IResourceContext } from "../../types/interfaces/IResourceContext";
import { ResourceType } from "../../types/enums/ResourceTypeEnum";
import { fetchResource } from "../../services/swapi";

export const ResourceContext = createContext<IResourceContext | null>(null);
