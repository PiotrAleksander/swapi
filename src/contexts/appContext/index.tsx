import React, {
  createContext,
  ReactNode,
  useState,
  useRef,
  useEffect,
  useCallback,
  Context
} from "react";
import { ResourceType } from "../../types/enums/ResourceTypeEnum";
import { IAppContext } from "../../types/interfaces/IAppContext";
import { fetchResources, setWinnerAndCounter } from "./utils";

const initialAppContext: IAppContext = {
  resourceType: ResourceType.PEOPLE,
  setResourceType: () => {},
  resources: [],
  counter: [0, 0],
  setCounter: () => {},
  onShuffle: () => {},
  isSending: false
};

let AppContext: Context<IAppContext>;
const { Provider, Consumer: AppConsumer } = (AppContext = createContext<
  IAppContext
>(initialAppContext));

const AppProvider = ({ children }: { children: ReactNode }) => {
  const [resourceType, setResourceType] = useState(ResourceType.PEOPLE);
  const [resources, setResources] = useState();
  const [counter, setCounter] = useState([0, 0]);
  const [isSending, setIsSending] = useState(false);
  const isMounted = useRef(true);

  useEffect(() => {
    return () => {
      // set unmounted on cleanup
      isMounted.current = false;
    };
  }, []);

  const onShuffle = useCallback(async () => {
    // don't send if sending already in progress
    if (isSending) return;
    setIsSending(true);
    try {
      // send the request
      const fetchedResources = await fetchResources(resourceType);

      const [firstResource, secondResource] = setWinnerAndCounter(
        fetchedResources,
        setCounter
      );

      setResources([firstResource, secondResource]);
    } catch (error) {
      setResources([]);
    } finally {
      if (isMounted.current)
        // only update if the provider is still mounted
        setIsSending(false);
    }
  }, [isSending, resourceType]); // update the callback if sending or resource type changes

  return (
    <Provider
      value={{
        resourceType,
        setResourceType,
        resources,
        counter,
        setCounter,
        onShuffle,
        isSending
      }}
    >
      {children}
    </Provider>
  );
};

export { AppProvider, AppConsumer };
export default AppContext;
