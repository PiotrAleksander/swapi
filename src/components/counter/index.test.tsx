import React from "react";
import { render, fireEvent, waitForElement } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import Counter from ".";
import { AppProvider, AppConsumer } from "../../contexts/appContext";

describe("Counter", () => {
  it("should render correctly", () => {
    const { container } = render(<Counter />);
    expect(container).toMatchSnapshot();
  });

  it("should render default counter value", () => {
    const { getAllByRole } = render(<Counter />);

    const counter = getAllByRole("heading");

    expect(counter[1]).toHaveTextContent("0 : 0");
  });

  it("should rerender correctly on counter change", async () => {
    const { getAllByRole, getByRole } = render(
      <AppProvider>
        <AppConsumer>
          {({ setCounter }) => (
            <button onClick={() => setCounter([1, 1])}></button>
          )}
        </AppConsumer>
        <Counter />
      </AppProvider>
    );

    expect(getAllByRole("heading")[1]).toHaveTextContent("0 : 0");

    fireEvent.click(getByRole("button"));

    const counter = await waitForElement(() => getAllByRole("heading")[1]);

    expect(counter).toHaveTextContent("1 : 1");
  });
});
