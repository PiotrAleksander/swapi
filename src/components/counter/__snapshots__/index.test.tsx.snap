// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter should render correctly 1`] = `
<div>
  <h4
    class="MuiTypography-root MuiTypography-body1"
  >
    Counter
  </h4>
  <h5
    class="MuiTypography-root MuiTypography-body1"
  >
    0 : 0
  </h5>
</div>
`;
